net: lwm2m: support for Block-Wise composite write operations

walzsi opened the pull request:
Currently only blockwise operations work for single resources, but with a few changes a composite write with several blocks is also possible via the TLV content format. This makes it possible, for example, to transfer certificates during the bootstrap.

rlubos commented on subsys/net/lib/lwm2m/lwm2m_message_handling.c:
```diff
@@ -175,8 +175,10 @@ static int get_block_ctx(const struct lwm2m_obj_path *path, struct lwm2m_block_c
-			  sizeof(struct lwm2m_obj_path)) == 0) {
+		if (path->level == block1_contexts[i].path.level &&
+		    path->obj_id == block1_contexts[i].path.obj_id &&
+		    (path->level < 2 || path->obj_inst_id == block1_contexts[i].path.obj_inst_id) &&
```
nit: Could use dedicated symbols (`LWM2M_PATH_LEVEL_OBJECT_INST`, `LWM2M_PATH_LEVEL_RESOURCE`) instad of magic numbers

SeppoTakalo replied:
Is this comparison leading to same result as using `lwm2m_obj_path_equal(path, block_contextx.path)`

See lwm2m_util.c

rlubos commented on subsys/net/lib/lwm2m/lwm2m_rw_oma_tlv.c:
```diff
@@ -959,9 +959,11 @@ int do_write_op_tlv(struct lwm2m_message *msg)
-	/* In case of block transfer go directly to the
-	 * message processing - consecutive blocks will not carry the TLV
-	 * header.
+	/* in case of block transfer, check if there are any fragments
```
`in` -> `In`

rlubos commented on subsys/net/lib/lwm2m/lwm2m_rw_oma_tlv.c:
```diff
@@ -971,7 +973,6 @@ int do_write_op_tlv(struct lwm2m_message *msg)
 			return ret;
 		}
 
-		return 0;
```
Just trying to understand if it's ok to remove this return here. What happens in case of single-resource block writes? After processing the message content, we'll proceed to the loop below, and what then? `oma_tlv_get()` is expected to fail?

walzsi replied:
exactly, in that case it will fail and break the loop.

walzsi committed:
net: lwm2m: check lwm2m path level while getting the block ctx

The Block1 context for composite operations is not found, as all path
levels are compared. The incoming path level should therefore always be
taken into account.

Signed-off-by: Simon Walz <simon.walz@autosen.com>

walzsi committed:
net: lwm2m: check each block for TLV header

For composite operations, we should always check whether the TLV header
is present. Otherwise, all blocks are assigned to one resource.

Signed-off-by: Simon Walz <simon.walz@autosen.com>

SeppoTakalo commented:
![image](https://github.com/user-attachments/assets/2b62943b-552d-406d-b337-594f193585f3)

Existing tests seem to pass OK.
